doc/source/data/examples/custom_datasource_example.rst

```diff
@@ -0,0 +1,107 @@
+Read and Write Custom File Types 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This guide shows you how to extend Ray Data to read and write file types that aren't 
+natively supported. Images are already supported with the :func:`~ray.data.read_images` 
```

Comment:
I think we need some disclaimers here, given that these examples currently require access to Data-internal APIs.

1. This is an advanced user / Data developer guide. We will be using Data-internal APIs that may not be as stable as public APIs.
2. map for read / and map_batches for write may be sufficient. Explain when you would want to use that vs. following this guide.

Reply:
Added disclaimers that this is an advanced guide and when to use `map` for reads. For writes, you'd always want to create a `Datasink`, since you can't generate filenames correctly with `map_batches`.